scipy/sparse/csgraph/_disjoint_set.pyx

```diff
@@ -0,0 +1,130 @@
+"""
+Disjoint set data structure
+"""
+
+import numpy as np
+cimport numpy as np
+cimport cython
+
+
+cdef class DisjointSet:
+    """Disjoint set data structure [1]_.
+
+    Parameters
+    ----------
+    n : int
+        The number of elements in the set.
+
+    Notes
+    -----
+    .. versionadded: 1.6.0
+
+
+    References
+    ----------
+    .. [1] https://en.wikipedia.org/wiki/Disjoint-set_data_structure
+
+
+    Examples
+    --------
+    >>> from scipy.sparse.csgraph import DisjointSet
+
+    Initialize a disjoint set with 4 elements:
+
+    >>> dis = DisjointSet(4)
+
+    Merge some subsets:
+
+    >>> dis.union(0, 1)
+    True
+    >>> dis.union(2, 3)
+    True
+    >>> dis.union(3, 3)
+    False
+
+    Find a root element:
+
+    >>> dis.find(1)
+    0
+
+    Attributes
+    ----------
+    n : int
+        The number of elements in the set.
+    nc : int
+        The number of components/subsets.
+
+    Methods
+    -------
+    union
+    find
+    """
+    cdef:
+        readonly np.npy_intp n
+        readonly np.npy_intp nc
+        readonly np.npy_intp[:] _parents
+        readonly np.npy_intp[:] _sizes
+
+    def __init__(DisjointSet self, np.intp_t n):
+        self.n = n
+        self.nc = n
+        self._sizes = np.ones(n, dtype=np.intp)
+        self._parents = np.arange(n, dtype=np.intp)
```

Comment:
Maybe this is a v2 request, but the python disjointsets doesn't required you to know the number of nodes (it internally uses a dict).

This means you can use arbitrary indexes (e.g. coords in a n'd graph) and avoids the requirement of knowing up front how many items you'll want to add.

Reply:
Thanks for bringing this up. I did consider using a dict-based design as it offers more flexibility. I decided against it though because of the performance cost of using dicts rather than integer arrays.

The other argument for integer arrays is that functions in `sparse.csgraph` operate on fixed-size graphs with integer-indexed nodes, so I think the current design is more appropriate for the module.